agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,563 @@ package gateways
+	"github.com/hashicorp/consul/agent/consul/fsm"
 	"github.com/hashicorp/consul/agent/consul/state"
 	"github.com/hashicorp/consul/agent/consul/stream"
 	"github.com/hashicorp/consul/agent/structs"
-	"github.com/hashicorp/go-hclog"
-	"github.com/pkg/errors"
 )
 
+type Updater struct {
+	UpdateWithStatus func(entry structs.ControlledConfigEntry) error
+	Update           func(entry structs.ConfigEntry) error
+	Delete           func(entry structs.ConfigEntry) error
+}
+
 type apiGatewayReconciler struct {
-	logger hclog.Logger
-	store  DataStore
+	fsm        *fsm.FSM
+	logger     hclog.Logger
+	updater    *Updater
+	controller controller.Controller
 }
 
-// NewAPIGatewayController returns a new APIGateway controller
-func NewAPIGatewayController(store DataStore, publisher state.EventPublisher, logger hclog.Logger) controller.Controller {
-	reconciler := apiGatewayReconciler{
-		logger: logger,
-		store:  store,
+func (r apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+	// We do this in a single threaded way to avoid race conditions around setting
+	// shared state. In our current out-of-repo code, this is handled via a global
+	// lock on our shared store, but this makes it so we don't have to deal with lock
+	// contention, and instead just work with a single control loop.
+	switch req.Kind {
+	case structs.APIGateway:
+		return reconcileEntry(r.fsm.State(), ctx, req, r.reconcileGateway, r.cleanupGateway)
+	case structs.BoundAPIGateway:
+		return reconcileEntry(r.fsm.State(), ctx, req, r.reconcileBoundGateway, r.cleanupBoundGateway)
+	case structs.HTTPRoute:
+		return reconcileEntry(r.fsm.State(), ctx, req, func(ctx context.Context, req controller.Request, store *state.Store, route *structs.HTTPRouteConfigEntry) error {
+			return r.reconcileRoute(ctx, req, store, route)
+		}, r.cleanupRoute)
+	case structs.TCPRoute:
+		return reconcileEntry(r.fsm.State(), ctx, req, func(ctx context.Context, req controller.Request, store *state.Store, route *structs.TCPRouteConfigEntry) error {
+			return r.reconcileRoute(ctx, req, store, route)
+		}, r.cleanupRoute)
+	case structs.InlineCertificate:
+		return r.enqueueCertificateReferencedGateways(r.fsm.State(), ctx, req)
+	default:
+		return nil
 	}
-	return controller.New(publisher, &reconciler).Subscribe(
-		&stream.SubscribeRequest{
-			Topic:   state.EventTopicAPIGateway,
-			Subject: stream.SubjectWildcard,
-		},
-	)
 }
 
-// Reconcile takes in a controller request and ensures this api gateways corresponding BoundAPIGateway exists and is
-// up to date
-func (r *apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+func reconcileEntry[T structs.ControlledConfigEntry](store *state.Store, ctx context.Context, req controller.Request, reconciler func(ctx context.Context, req controller.Request, store *state.Store, entry T) error, cleaner func(ctx context.Context, req controller.Request, store *state.Store) error) error {
```

Comment:
This does the initial fetch operation for a config entry and then hands off to either a cleanup function or a reconciliation function in the case of this being an update or create operation.